File: 5_stone_game/stone_game.go
// Code written for the Stone Game problem from Hackerrank
// Go version
package main 

import "fmt"

func p2 (c int, moves [3]int) int{
	min_result := 1;
	for i:=0; i<len(moves); i++{
		if c - moves[i] >= 0{
			result := p1(c-moves[i], moves)
			if result < min_result{
				min_result = result;
			}
		}
	}
	return min_result
}

func p1(c int, moves [3]int) int{
	max_result := -1;
	for i:=0; i<len(moves); i++{
		if c - moves[i] >= 0{
			result := p2(c-moves[i], moves)
			if result > max_result{
				max_result = result;
			}
		}
	}
	return max_result
}

func main() {
	var cases = [8]int{1,2,3,4,5,6,7,10}
	var moves = [3]int{2,3,5};
	for i:=0; i<len(cases); i++{
		outcome := p1(cases[i], moves)
		fmt.Println(outcome)
	}
}

File: 5_stone_game/stone_game.java
public class stone_game {
  public static void main(String... args){
    int[] cases = {1,2,3,4,5,6,7,10};
    int[] moves = {2,3,5};
    String[] convert = {"Second", "First"};
    game(cases, convert, moves);
  }

  private static int p2(int c, int[] moves){
    int minResult = 1;
    for(int m: moves){
      if(c - m >= 0){
        int result = p1(c-m, moves);
        if(result < minResult)
          minResult = result;
      }
    }
    return minResult;
  }

  private static int p1(int c, int[] moves){
    int maxResult = 0;
    for(int m: moves){
      if(c - m >= 0){
        int result = p2(c - m, moves);
        if(result > maxResult)
          maxResult = result;
      }
    }
    return maxResult;
  }

  private static void game(int[] cases, String[] convert, int[] moves){
    for(int c: cases){
      int outcome = p1(c, moves);
      System.out.println(convert[outcome]);
    }
  }
}
